### ISSUE TITLE
NVIDIA driver 460.79 blocks reading GPU power consumption?

### ISSUE THREAD

**Issue by shocker-0x15** [openhardwaremonitor/openhardwaremonitor#1376]
Hi,
Since I updated the NVIDIA driver to 460.79, Open Hardware Monitor seems not able to show GPU power consumption.
I searched to see others meeting the same issue, there seems:
https://forums.guru3d.com/threads/nvidia-geforce-460-79-driver-download-and-discussion.435609/page-4

Open Hardware Monitor 0.9.5
Windows 10 20H2
Core i9-9900K
Geforce RTX 3080
NVIDIA Driver 460.79

Thanks,
[OpenHardwareMonitor.Report.txt](https://github.com/openhardwaremonitor/openhardwaremonitor/files/5683601/OpenHardwareMonitor.Report.txt)

**Comment by MerlinStWahgwaan** [comment]
I am having the same issue with OHM...

almost same specs too...

i9-9900K
RTX2080
Nvidia Driver 460.79

Worked fine yesterday, did the update last night, and now today, no GPU Power readings....

**Comment by Zefod** [comment]
Same here;  it's probably because of the latest NVIDIA driver... 

Windows 10 20H2
Core  i7-4790
Geforce RTX 2080 Super
NVIDIA Driver 460.79

**Comment by woahwotsdis** [comment]
Having the same issue but with a 10 series card and tried both of the NVIDIA driver updates (460.79 & 460.89)

Windows 10 20H2
R5-1600AF 
Geforce RTX 1080
NVIDIA Driver 460.89

**Comment by taleshape** [comment]
I have the same issue.

Windows 10 Pro 20H2
AMD Ryzen 9 3900X
GeForce RTX 2060 SUPER
NVIDIA Drivers 460.79 & 460.89

**Comment by anjisZhou** [comment]
`/*
 
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.
 
  Copyright (C) 2020 Michael Möller <mmoeller@openhardwaremonitor.org>
 
*/

using System;
using System.Runtime.InteropServices;

namespace OpenHardwareMonitor.Hardware.Nvidia {
  internal class NVML {
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate NvmlReturn nvmlInitWithFlagsDelegate(uint flag);
    private static readonly nvmlInitWithFlagsDelegate nvmlInitWithFlags;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate NvmlReturn nvmlInitDelegate();
    private static readonly nvmlInitDelegate nvmlInit;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    private delegate NvmlReturn nvmlInit_v2Delegate();
    private static readonly nvmlInit_v2Delegate nvmlInit_v2;
      
    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate NvmlReturn nvmlShutdownDelegate();
    public static readonly nvmlShutdownDelegate NvmlShutdown;

    public delegate NvmlReturn nvmlDeviceGetHandleByPciBusIdDelegate(
      string pciBusId, out NvmlDevice device);
    public static readonly nvmlDeviceGetHandleByPciBusIdDelegate
      NvmlDeviceGetHandleByPciBusId;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate NvmlReturn nvmlDeviceGetPowerUsageDelegate(
      NvmlDevice device, out int power);
    public static readonly nvmlDeviceGetPowerUsageDelegate 
      NvmlDeviceGetPowerUsage;

    [UnmanagedFunctionPointer(CallingConvention.Cdecl)]
    public delegate NvmlReturn nvmlDeviceGetPcieThroughputDelegate(
      NvmlDevice devicee, NvmlPcieUtilCounter counter, out uint value);
    public static readonly nvmlDeviceGetPcieThroughputDelegate
      NvmlDeviceGetPcieThroughput;


    public static NvmlReturn NvmlInit() {
      try {
        var result = nvmlInitWithFlags(1);
        //var result = nvmlInit_v2();
        initialized = result == NvmlReturn.Success || result == NvmlReturn.ErrorAlreadyInitialized;
        return result;
      } catch {
        try {
          var result = nvmlInit();
          initialized = result == NvmlReturn.Success;
          return result;
        } catch {
          return NvmlReturn.ErrorLibraryNotFound;
        }
      }
    }

    private static string GetDllName() {
      if (OperatingSystem.IsUnix)
        return "libnvidia-ml.so";
      else
        return "nvml.dll";
    }

    private static T CreateDelegate<T>(string entryPoint) where T : Delegate {
      var attribute = new DllImportAttribute(GetDllName()) {
        CallingConvention = CallingConvention.Cdecl,
        PreserveSig = true,
        EntryPoint = entryPoint
      };
      PInvokeDelegateFactory.CreateDelegate(attribute, out T newDelegate);
      return newDelegate;
    }

    static NVML() {
      nvmlInitWithFlags =
  CreateDelegate<nvmlInitWithFlagsDelegate>(
  "nvmlInitWithFlags");
      nvmlInit =
        CreateDelegate<nvmlInitDelegate>(
        "nvmlInit");
      nvmlInit_v2 =
        CreateDelegate<nvmlInit_v2Delegate>(
        "nvmlInit_v2");
      NvmlShutdown =
        CreateDelegate<nvmlShutdownDelegate>(
        "nvmlShutdown");
      NvmlDeviceGetHandleByPciBusId =
        CreateDelegate<nvmlDeviceGetHandleByPciBusIdDelegate>(
        "nvmlDeviceGetHandleByPciBusId_v2");
      NvmlDeviceGetPowerUsage =
        CreateDelegate<nvmlDeviceGetPowerUsageDelegate>(
        "nvmlDeviceGetPowerUsage");
      NvmlDeviceGetPcieThroughput =
        CreateDelegate<nvmlDeviceGetPcieThroughputDelegate>(
        "nvmlDeviceGetPcieThroughput");
    }

    private static bool initialized;
    public static bool IsInitialized {
      get { return initialized; }
    }

    [StructLayout(LayoutKind.Sequential)]
    internal struct NvmlDevice {
      private readonly IntPtr ptr;      
    }

    internal enum NvmlPcieUtilCounter {
      TxBytes = 0,
      RxBytes = 1
    }

    internal enum NvmlReturn {
      /// <summary>
      /// The operation was successful
      /// </summary>
      Success = 0,

      /// <summary>
      /// NVML was not first initialized with nvmlInit()
      /// </summary>
      ErrorUninitialized = 1,

      /// <summary>
      /// A supplied argument is invalid
      /// </summary>
      ErrorInvalidArgument = 2,

      /// <summary>
      /// The requested operation is not available on target device
      /// </summary>
      ErrorNotSupported = 3,

      /// <summary>
      /// The current user does not have permission for operation
      /// </summary>
      ErrorNoPermission = 4,

      /// <summary>
      /// Deprecated: Multiple initializations are now allowed through ref 
      /// counting
      /// </summary>
      ErrorAlreadyInitialized = 5,

      /// <summary>
      /// A query to find an object was unsuccessful
      /// </summary>
      ErrorNotFound = 6,

      /// <summary>
      /// An input argument is not large enough
      /// </summary>
      ErrorInsufficientSize = 7,

      /// <summary>
      /// A device's external power cables are not properly attached
      /// </summary>
      ErrorInsufficientPower = 8,

      /// <summary>
      /// NVIDIA driver is not loaded
      /// </summary>
      ErrorDriverNotLoaded = 9,

      /// <summary>
      /// User provided timeout passed
      /// </summary>
      ErrorTimeout = 10,

      /// <summary>
      /// NVIDIA Kernel detected an interrupt issue with a GPU
      /// </summary>
      ErrorIrqIssue = 11,

      /// <summary>
      /// NVML Shared Library couldn't be found or loaded
      /// </summary>
      ErrorLibraryNotFound = 12,

      /// <summary>
      /// Local version of NVML doesn't implement this function
      /// </summary>
      ErrorFunctionNotFound = 13,

      /// <summary>
      /// infoROM is corruptedinfoROM is corrupted
      /// </summary>
      ErrorCorruptedInforom = 14,

      /// <summary>
      /// The GPU has fallen off the bus or has otherwise become inaccessible
      /// </summary>
      ErrorGpuIsLost = 15,

      /// <summary>
      /// The GPU requires a reset before it can be used again
      /// </summary>
      ErrorResetRequired = 16,

      /// <summary>
      /// The GPU control device has been blocked by the operating 
      /// system/cgroups
      /// </summary>
      ErrorOperatingSystem = 17,

      /// <summary>
      /// RM detects a driver/library version mismatch
      /// </summary>
      ErrorLibRmVersionMismatch = 18,

      /// <summary>
      /// An operation cannot be performed because the GPU is currently in use
      /// </summary>
      ErrorInUse = 19,

      /// <summary>
      /// Insufficient memory
      /// </summary>
      ErrorMemory = 20,

      /// <summary>
      /// No data
      /// </summary>
      ErrorNoData = 21,

      /// <summary>
      /// The requested vgpu operation is not available on target device, 
      /// becasue ECC is enabled
      /// </summary>
      ErrorvGpuEccNotSupported = 22,

      /// <summary>
      /// An internal driver error occurred
      /// </summary>
      ErrorUnknown = 999
    };

  }
}
`
To replace NVML.cs, it can be solved

**Comment by MerlinStWahgwaan** [comment]
Working HOTFIX from Nvidia....
Must be installed manually, but I can fonfirm, it works!

https://www.nvidia.com/en-us/geforce/forums/game-ready-drivers/13/421696/announcing-geforce-hotfix-driver-46097-released-12/

**Comment by rogueblue42** [comment]
> 
> 
> Working HOTFIX from Nvidia....
> Must be installed manually, but I can fonfirm, it works!
> 
> https://www.nvidia.com/en-us/geforce/forums/game-ready-drivers/13/421696/announcing-geforce-hotfix-driver-46097-released-12/

Thank you!!! This got the power back :-)

**Comment by GettnBetter** [comment]
I did not have the same experience.  The hotfix and all future releases of the NVIDIA drivers seem to block this ability.  I don't get any power readings from my 2070 Super anymore.

**Comment by MerlinStWahgwaan** [comment]
Thats strange... I also have a 2070 Super in my desktop, and its been
working fine since the hotfix and the new 461.09 driver....

Make sure OHM is fully closed when you install the driver, or restart it
after installing the driver.

Other than that, I would hope someone else has an idea for you because it
should work now...

**Comment by MerlinStWahgwaan** [comment]
Also, check in OHM directly, and make sure its not just your gadget not showing the readings, if thats what your using... it might recognize it as a new reading and need to be re added to the gadget

**Comment by rogueblue42** [comment]
I can confirm that the latest Nvidia release 461.09 still works for my 3080, just as the hot fix did.

**Comment by MerlinStWahgwaan** [comment]
> I did not have the same experience. The hotfix and all future releases of the NVIDIA drivers seem to block this ability. I don't get any power readings from my 2070 Super anymore.

Any luck?

**Comment by GettnBetter** [comment]
Thank you for the suggestions!   Very much appreciated.

I closed OHM and then manually installed the latest NVIDIA drivers and all is working.

I have power again woot!   The PCIe bandwidth came back too.

**Comment by MerlinStWahgwaan** [comment]
Awsome. Glad to hear it!
Have fun and enjoy!

**Comment by shocker-0x15** [comment]
I had been not able to get back power reading even if reinstall the driver.
But I tried again to install the latest driver 461.09 while ensuring closing the OHM before install and selected clean install to make sure. I've confirmed the OHM now has power reading.

I'm closing this thread.
Thanks for cooperation!

**Comment by rogueblue42** [comment]
Not sure if I can still respond to this but this issue is back for me 461.09 even though it worked before. I used DDU to do a clean wipe of the graphic driver and reinstalled 461.09 and the GPU power is no longer displayed.

**Comment by MerlinStWahgwaan** [comment]
> Not sure if I can still respond to this but this issue is back for me 461.09 even though it worked before. I used DDU to do a clean wipe of the graphic driver and reinstalled 461.09 and the GPU power is no longer displayed.

Have you followed the above reinstall suggestion, make sure OHM is fully closed, fresh reinstall of drivers, then reopen OHM?
That seems to have fixed everyone's issues thus far.
Doesn't hurt to add in a reboot before relaunching OHM, just to ensure the new drivers are in proper use....

I have not had an issues with the current drivers, on my 2080 or my 2070 Super computers. In fact, I had a separate issue wirh other hardware just last week, that ended up forcing me to do a complete factory restore on my Alienware with the 2080, back to fresh windows, everything... stupid but I had no choice because I couldn't solve my issue... and since updating back to most up-to-date, still no issues with OHM or power readings from GPU, except in Intel XTU, it seems unable to grab proper GPU data since the 460.79 update that started this issue for everyone on OHM...... which I've raised to Nvidia myself as an issue but they don't seem in too much of a rush to fix it....

If you try the fully closed OHM and GPU driver reinstall and that still fails, I would create a separate thread for your current issue with the current driver number listed, maybe someone will have other ideas for you....

**Comment by MerlinStWahgwaan** [comment]
> Not sure if I can still respond to this but this issue is back for me 461.09 even though it worked before. I used DDU to do a clean wipe of the graphic driver and reinstalled 461.09 and the GPU power is no longer displayed.

Double check your driver with Nvidia website....
I just double checked mine because your version sounded odd to me....

My current driver on both my devices, is 461.40

You said you have 461.09.

You should close OHM, and update to the newest driver manually if GeForce Experience is not finding it automatically for you.... but check the Nvidia website for your specific device and make sure your on the most recent driver forsure before relaunching OHM

**Comment by rogueblue42** [comment]
> 
> 
> > Not sure if I can still respond to this but this issue is back for me 461.09 even though it worked before. I used DDU to do a clean wipe of the graphic driver and reinstalled 461.09 and the GPU power is no longer displayed.
> 
> Double check your driver with Nvidia website....
> I just double checked mine because your version sounded odd to me....
> 
> My current driver on both my devices, is 461.40
> 
> You said you have 461.09.
> 
> You should close OHM, and update to the newest driver manually if GeForce Experience is not finding it automatically for you.... but check the Nvidia website for your specific device and make sure your on the most recent driver forsure before relaunching OHM

Thank you, this worked. I updated Nvidia directly from 461.09 to 461.72. I closed OHM, did a clean install, and reboot. OHM again showed power.
